conn.go

```diff
@@ -791,60 +750,137 @@ func (c *Conn) isHandshakeCompletedSuccessfully() bool {
 	return boolean.bool
 }
 
-func (c *Conn) startHandshakeOutbound() {
+func (c *Conn) handshake(ctx context.Context, cfg *handshakeConfig, initialFlight flightVal, initialState handshakeState) error {
+	c.fsm = newHandshakeFSM(&c.state, c.handshakeCache, cfg, initialFlight)
+
+	done := make(chan struct{})
+	ctxRead, cancelRead := context.WithCancel(context.Background())
+	c.cancelHandshakeReader = cancelRead
+	cfg.onFlightState = func(f flightVal, s handshakeState) {
+		if s == handshakeFinished && !c.isHandshakeCompletedSuccessfully() {
+			c.setHandshakeCompletedSuccessfully()
+			close(done)
+		}
+	}
+
+	ctxHs, cancel := context.WithCancel(context.Background())
+	c.cancelHandshaker = cancel
+
+	firstErr := make(chan error, 1)
+
+	c.wg.Add(2)
+
+	// Handshake routine should be live until close.
+	// The other party may request retransmission of the last flight to cope with packet drop.
 	go func() {
-		defer func() {
-			if c.handshakeErr.load() != nil {
-				if err := c.close(); err != nil {
-					c.log.Errorf(fmt.Sprintf("Failed to close (%v)", err))
-				}
-			}
-		}()
-		for {
-			var (
-				isFinished bool
-				alertPtr   *alert
-				err        error
-			)
+		defer c.wg.Done()
+		err := c.fsm.Run(ctxHs, c, initialState)
+		if err != context.Canceled {
 			select {
-			case <-c.handshakeDoneSignal.Done():
-				return
-			case <-c.workerTicker.C:
-				isFinished, alertPtr, err = c.flightHandler(c)
-			case <-c.currFlight.workerTrigger:
-				isFinished, alertPtr, err = c.flightHandler(c)
+			case firstErr <- err:
+			default:
 			}
-
-			if alertPtr != nil {
-				if alertErr := c.notify(alertPtr.alertLevel, alertPtr.alertDescription); alertErr != nil {
-					err = fmt.Errorf("%v %v", err, alertErr)
+		}
+	}()
+	go func() {
+		defer c.wg.Done()
+		for {
+			if err := c.readAndBuffer(ctxRead); err != nil {
+				switch e := err.(type) {
```

Comment:
Not a big fan of unsetting a value on the stack. Would you mind refactoring this so it is more explicit? Maybe early returns are better?

Reply:
Updated to use `continue` instead of `err = nil`.